src/cluster.c

```diff
@@ -4948,6 +4962,163 @@ void addNodeReplyForClusterSlot(client *c, clusterNode *node, int start_slot, in
+     * 6. replication offset (in case of replica only).
+     * 7. health
+     */
+    int reply_cnt = 0;
+    void *node_replylen = addReplyDeferredLen(c);
+    addReplyBulkCString(c, "id");
+    addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
+    reply_cnt++;
+    addReplyBulkCString(c, "port");
+    addReplyLongLong(c, node->port);
+    reply_cnt++;
+    addReplyBulkCString(c, "ip");
+    addReplyBulkCString(c, node->ip);
+    reply_cnt++;
+    if (node->hostname) {
+        addReplyBulkCString(c, "hostname");
+        addReplyBulkCString(c, node->hostname);
+        reply_cnt++;
+    }
+    const char *endpoint = getPreferredEndpoint(node);
+    if (!strcmp(endpoint, "")) {
+        addReplyBulkCString(c, "endpoint");
+        addReplyBulkCString(c, endpoint);
+        reply_cnt++;
+    }
+
+    addReplyBulkCString(c, "role");
+    if (nodeIsSlave(node)) {
+        addReplyBulkCString(c, "replica");
+        addReplyBulkCString(c, "replication-offset");
+        addReplyLongLong(c, node->repl_offset);
+        reply_cnt+=2;
+    } else {
+        addReplyBulkCString(c, "primary");
+        reply_cnt++;
+    }
+
+    addReplyBulkCString(c, "health");
+    const char *health_msg = NULL;
+    if (nodeFailed(node)) {
+        health_msg = "FAIL";
+    } else if (nodeIsSlave(node) && node->repl_offset == 0) {
+        health_msg = "LOADING";
+    } else {
+        health_msg = "ONLINE";
+    }
+    addReplyBulkCString(c, health_msg);
+    reply_cnt++;
+
+    setDeferredMapLen(c, node_replylen, reply_cnt);
+}
+
+void addNodeReplyForClusterShard(client *c, clusterNode *node, list *slotInfo) {
+    addReplyMapLen(c, 2);
+    addReplyBulkCString(c,"slots");
+    uint16_t slotInfoCount = 0;
+    if (slotInfo != NULL) {
+        slotInfoCount = listLength(slotInfo);
+        addReplyArrayLen(c, slotInfoCount);
+        listIter li;
+        listRewind(slotInfo, &li);
+        listNode *ln;
+        while((ln = listNext(&li))) {
+            addReplyBulkSds(c, listNodeValue(ln));
+        }
+    } else {
+        addReplyArrayLen(c, slotInfoCount);
+    }
+    addReplyBulkCString(c,"nodes");
+    list *nodes_for_slot = clusterGetNodesServingMySlots(node);
+    if (nodes_for_slot != NULL) {
+        addReplyArrayLen(c, listLength(nodes_for_slot));
+        if (listLength(nodes_for_slot) != 0) {
+            listIter li;
+            listNode *ln;
+            listRewind(nodes_for_slot, &li);
+            while ((ln = listNext(&li))) {
+                clusterNode *node = listNodeValue(ln);
+                addNodeDetailsToShardReply(c, node);
+            }
+            listRelease(nodes_for_slot);
+        }
+    } else {
+        addReplyArrayLen(c, 0);
+    }
+}
+
+
+void clusterReplyShards(client *c) {
+    clusterNode *n = NULL;
+    int start = -1;
+    void *shard_replylen = addReplyDeferredLen(c);
+    int shard_cnt = 0;
+    dict *nodeToSlotPair = dictCreate(&clusterNodesToSlotDictType);
+    for (int i = 0; i <= CLUSTER_SLOTS; i++) {
```

Comment:
I've thought about how to deduplicate this slot range enumeration logic too. My idea is create an iterator function which returns each slot range of a node:

```c
int start = 0, end = 0;
while (nextSlotRange(node, &start, &end)) {
   // do something for each range
}
```

I have an implementation. I can post it tomorrow.

Reply:
Worth mentioning we have two types of iterations:
1. We do a straight iteration through all of the slots and group them by node (see `CLUSTER SLOTS`)
2. We want a result which is the ordered list of nodes and then we want to iterate through their slots (see `CLUSTER NODES`. 

Need to make sure the iterator supports both, or still makes 2 easily attainable.

Reply:
@madolson Good point. The version posted below only iterates over a single node (i.e. it covers 2 but not 1) but I guess it can be adapted, or possibly 1 is different enough to have a separate iteration function for that.

<details>
<summary>int nextSlotRange(clusterNode *node, int *begin, int *end);</summary>

```C
/* Iterate over the slot ranges of a node. Use like this:
 *
 *     int start = 0, end = 0;
 *     while (nextSlotRange(node, &start, &end) {
 *         // This node has range start..end.
 *     }
 */
int nextSlotRange(clusterNode *node, int *begin, int *end) {
    int j;
    int start = -1;
    for (j = (*end == 0) ? 0 : *end + 1; j < CLUSTER_SLOTS; j++) {
        int bit = clusterNodeGetSlotBit(node, j);
        if (bit && start == -1) start = j;
        if (start != -1 && (!bit || j == CLUSTER_SLOTS-1)) {
            if (bit && j == CLUSTER_SLOTS-1) j++;
            *begin = start;
            *end = j - 1;
            return 1;
        }
    }
    return 0;
}
```
</details>

Reply:
@zuiderkwast  I have added a conditional check if the node provided is null, we move on to the next slot. WDYT ?

```
int nextSlotRange(clusterNode *node, int *begin, int *end) {
    int j;
    int start = -1;
    if (node == NULL) {
        *begin = *begin + 1;
        *end = *begin;
        if (*end == CLUSTER_SLOTS) {
            return 0;
        } else {
            return 1;
        }
    }
    for (j = (*end == 0) ? 0 : *end + 1; j < CLUSTER_SLOTS; j++) {
        int bit = clusterNodeGetSlotBit(node, j);
        if (bit && start == -1) start = j;
        if (start != -1 && (!bit || j == CLUSTER_SLOTS - 1)) {
            if (bit && j == CLUSTER_SLOTS - 1) j++;
            *begin = start;
            *end = j - 1;
            return 1;
        }
    }
    return 0;
}
```

```
    int start = 0, end = 0;
    clusterNode *n = server.cluster->slots[start];
    while (nextSlotRange(n, &start, &end)) {
        if (n != NULL) {
            sds name = sdsnewlen(n->name, CLUSTER_NAMELEN);
            dictEntry *de = dictFind(nodeToSlotPair, name);
            sdsfree(name);
            list *slotPair;
            if (de == NULL) {
                slotPair = listCreate();
                dictAdd(nodeToSlotPair, sdsnewlen(n->name, CLUSTER_NAMELEN), slotPair);
            } else {
                slotPair = dictGetVal(de);
            }
            if (start != end) {
                listAddNodeTail(slotPair, sdscatfmt(sdsempty(), "%i-%i", start, end));
            } else {
                listAddNodeTail(slotPair, sdscatfmt(sdsempty(), "%i", start));
            }
        }
        if ((end + 1) != CLUSTER_SLOTS) {
            n = server.cluster->slots[end + 1];
        }
    }
```

Reply:
@hpatro if n == NULL, the function returns a range with a single slot for no node? I think it is strange. It relies on the caller to make sure that `begin + 1` has no node. Maybe you meant `*begin = *end + 1`? I think the next range starts where the previous range ends. Or did you change this?

Anyway, this function cannot check this slot has no node, so I think it is strange. Better check if n == NULL and find node with a slot before calling the iterator function.

Maybe we need another function `nextSlotRangeAndNode(*begin, *end, *node)` which also returns the node of the next range...? This logic is different and it is less complex to make them separate functions IMO.

```C
    int start = 0, end = 0;
    clusterNode *n;
    while (nextSlotRangeAndNode(&start, &end, &n)) {
        if (n == NULL) continue;
        // use node and range
    }
```

Implementation (untested):

```C
/* Finds the next slot range and its node, starting at end + 1. The returned node
 * is NULL if the slot range is unassigned. */
int nextSlotRangeAndNode(int *begin, *end, clusterNode **node) {
    int start = (*end == 0) ? 0 : *end + 1;
    if (start >= CLUSTER_SLOTS)
        return 0;
    clusterNode *n = server.cluster->slots[start];
    int j = start;
    while (j < CLUSTER_SLOTS && server.cluster->slots[j] == n)
        j++;
    *begin = start;
    *end = j - 1;
    return 1;
}
```

----

One question: Why do we need to return the slots in ascending order? Can't we just use the order of the master nodes?

Reply:
Probably fine, @zuiderkwast maybe make an issue for improvement?